### PR TITLE
1318 admin user tab

### PR DESCRIPTION
**Pull request by nickbristow**
## Background
As an OEC Admin User, I can view a table of all users, along with the organizations they're associated with.
## GitHub Issue
#1318

## Validation Plan
Verify as admin
verify regular user cannot access

## Automated Testing
- [x] All unit tests are passing.
- [x] All e2e tests are passing.
- [ ] If applicable, unit tests have been added to cover this changeset.
- [ ] If applicable, e2e tests have been added to cover this changeset.

## Additional Context
If there's anything additional information that those reviewing or validating this work should be aware of, please specify it here.

### CODE REVIEW

**Review by vonbearshark — APPROVED** [review]
lgtm

File: client/src/containers/Users/UsersTable.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  },
+  {
+    name: 'Email address',
+    className: 'text-no-wrap',
+    sort: (row) => row.email ?? '',
+    cell: ({ row }) => <td>{row.email ?? ''}</td>,
+  },
+  {
+    name: 'Organization(s)',
+    cell: ({ row }) => <td className="maxw-card-lg">{row.organizations}</td>,
```
[the remark this answers is not in the document]
Comment (nickbristow) [reply client/src/containers/Users/UsersTable.tsx:41]: done in the sql query

**Comment by nickbristow** [comment]
> I meant `2 for Unused variable, import, function or class`, in the LGTM bot, not the `local` tests

ah...yes. I glossed over it because i assumed.

File: client/src/containers/Users/Users.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { UsersTable } from './UsersTable';
+import { apiGet } from '../../utils/api';
+
+const Users: React.FC = () => {
+  const { user } = useContext(UserContext);
+  const h1Ref = getH1RefForTitle();
+  const { accessToken } = useContext(AuthenticationContext);
+
+  const [data, setData] = useState<UserSummary[]>([]);
+  apiGet(`users/`, accessToken).then((res: UserSummary[]) => setData(res));
```
Comment (bamader) [inline client/src/containers/Users/Users.tsx:16]: Nit: I think this should have an `await` in front of it or be inside a `useEffect` that returns if the access token or response isn't present yet; that's typically how we prevent weird flashing behavior as screens load and change.

**Review by bamader — APPROVED** [review]

File: client/src/containers/Users/Users.tsx (hunk 2 above)
Comment (vonbearshark) [reply client/src/containers/Users/Users.tsx]: Does the new code prevent the flashing? Not clear to me how..?

**Review by JordanGuinn — APPROVED** [review]

File: client/src/containers/Users/Users.tsx (hunk 2 above)
Comment (nickbristow) [reply client/src/containers/Users/Users.tsx]: @vonbearshark i think it would really just prevent it from calling the api before its ready.

File: client/src/containers/Users/Users.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  const { user } = useContext(UserContext);
+  const h1Ref = getH1RefForTitle();
+  const { accessToken } = useContext(AuthenticationContext);
+
+  const [data, setData] = useState<UserSummary[]>([]);
+
+  useEffect(() => {
+    (async function loadUsers() {
+      if (user && accessToken) {
+        await apiGet(`users/`, accessToken).then((res: UserSummary[]) =>
```
Comment (jlhogan) [inline client/src/containers/Users/Users.tsx:20]: shouldn't need the trailing slash, right?

File: client/src/containers/Users/UsersTable.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { Column, Table } from '@ctoec/component-library';
+import React from 'react';
+import { UserSummary } from '../../shared/payloads/UsersResponse';
+
+type UserTableProps = {
+  users?: Array<UserSummary>;
+};
+
+export const UsersTable: React.FC<UserTableProps> = ({ users = [] }) => (
+  <Table<UserSummary>
```
Comment (jlhogan) [inline client/src/containers/Users/UsersTable.tsx:10]: so this is something I just though of but might it be worth it to try pulling in the carbon data table component now instead of continuing to build more tables with our very complex and very bespoke table component?

File: client/src/containers/Users/UsersTable.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  />
+);
+
+const TableColumns: () => Column<UserSummary>[] = () => [
+  {
+    name: 'Name',
+    sort: (row) => row.name,
+    cell: ({ row }) => (
+      <th scope="row">
+        {/* Deactivate links until org detail pages are built */}
```
Comment (jlhogan) [inline client/src/containers/Users/UsersTable.tsx:25]: copy pasta comment

File: src/controllers/users.ts [hunk 6]
```diff
@@ -28,3 +28,12 @@ export const addDataToUser = async (user: User) => {
   user.sites = sites;
   user.organizations = organizations;
 };
+
+export const getUsers = async (): Promise<User[]> =>
+  getManager().query(
```
Comment (jlhogan) [inline src/controllers/users.ts:33]: can we try to keep these handwritten queries styled a bit better for readability? my pref:
```
SELECT
  u.id,
  CONCAT(u.lastName, ', ', u.firstName) as name,
  u.email,
  STRING_AGG(o.providerName,' ', ') as organizations
FROM [user] u 
JOIN organization_permission op on u.id = op.userId
JOIN organization o on op.organizationid = o.id
GROUP BY u.id, u.lastName, u.firstName, u.email
```

**Review by jlhogan — APPROVED** [review]

File: client/src/containers/Users/UsersTable.tsx (hunk 4 above)
Comment (JordanGuinn) [reply client/src/containers/Users/UsersTable.tsx:10]: **Fully** agreed, assuming this doesn't hugely bump the amount of effort of this ticket (which I wouldn't think would be the case, this is as a simple of a table as can be)

File: client/src/containers/Users/UsersTable.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
+// import { Column, Table } from '@ctoec/component-library';
```
Comment (JordanGuinn) [inline client/src/containers/Users/UsersTable.tsx:1]: nit: this can be deleted now

File: client/src/containers/Organizations/Organizations.tsx [hunk 8]
```diff
@@ -14,9 +14,17 @@ const Organizations: React.FC = () => {
   const { accessToken } = useContext(AuthenticationContext);
 
   const [data, setData] = useState<OrganizationSummary[]>([]);
-  apiGet(`organizations/`, accessToken).then((res: OrganizationSummary[]) =>
-    setData(res)
-  );
+
+  useEffect(() => {
+    (async function loadUsers() {
```
Comment (JordanGuinn) [inline client/src/containers/Organizations/Organizations.tsx:19]: 1.  Why this change here?
2.  Presumably this should be named `loadOrganizations`?

File: client/src/containers/Users/UsersTable.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  TableBody,
+  TableCell,
+} from 'carbon-components-react';
+
+type UserTableProps = {
+  users?: Array<UserSummary>;
+};
+
+export const UsersTable: React.FC<UserTableProps> = ({ users = [] }) => (
+  <Table className="oec-table oec-table--full-width">
```
Comment (JordanGuinn) [inline client/src/containers/Users/UsersTable.tsx:18]: Oh HELLO there, Carbon

**Review by JordanGuinn — APPROVED** [review]

File: client/src/containers/Organizations/Organizations.tsx (hunk 8 above)
Comment (nickbristow) [reply client/src/containers/Organizations/Organizations.tsx:19]: 1. @bamader pointed this out for the users page which is essentially the same as organizations. This prevents unnecessary reloading and only loads the data once its ready. 
2. Yes, updated name

**Review by bamader — APPROVED** [review]

**Comment by bamader** [comment]
Validated that this works. Regular users cannot access the admin/user page, and on the page, all users are displayed with their associated organizations. Since we don't yet have the functionality to create a user, that's basically all we have to check for this.
